oximeter/db/src/client.rs

```diff
@@ -79,18 +80,21 @@ impl Client {
         //  to/from the database, as well as the cost of parsing them for each measurement, only to
         //  promptly throw away almost all of them (except for the first).
         let timeseries_name = TimeseriesName::try_from(timeseries_name)?;
-        let schema = self
-            .schema_for_timeseries(&timeseries_name)
-            .await?
-            .ok_or_else(|| {
-                Error::QueryError(format!(
-                    "No such timeseries: '{}'",
-                    timeseries_name
-                ))
-            })?;
-        let mut query_builder = query::SelectQueryBuilder::new(&schema)
+        let schema = match self.schema_for_timeseries(&timeseries_name).await? {
+            Some(schema) => schema,
+            // If the timeseries doesn't exist, just return an empty Vec.
+            None => return Ok(Vec::new()),
```

Comment:
I'm not sure this is what we want to do if there's no timeseries with the provided name. This would be equivalent to querying in something like CRDB for a table that doesn't exist, whic seems like an error we want to propagate, no? I imagine you might have added this to support the tests, where we might be querying for data until it shows up. Is that right? IIRC, in the similar tests I had written, I inspected the error and ignored the particular variant this used to generate.

Reply:
So if I revert this to the prior implementation, then:

- When we query for a metric that *should exist* (it's in the API)...
- ... But no samples are recorded yet (we hit this case?)
- ... a 500 error is propagated upwards, with the internal message: `Invalid query or data filter: No such timeseries: 'crucible_upstairs:read'`.

Regardless of test behavior, this kinda seems wrong to me. If a user is querying for an endpoint that we validate in the API, it seems more correct to return a 200 result + empty vector than a 500 error.

If the layering is the issue, I'd be happy to match on this error in Nexus itself, but I wanted to mention why error matching in the tests felt odd.

Reply:
> If a user is querying for an endpoint that we validate in the API, it seems more correct to return a 200 result + empty vector than a 500 error.

I agree with this. I just think it should be Nexus that converts the database error into a 200 and empty vector. From the point of view of the ClickHouse client isolated, this change would mean that any non-existent timeseries returns a successful result. I don't see how you'd distinguish between the cases where there is no such timeseries (no schema with that name) and an extant timeseries that has no data in it.

Reply:
I did a couple things:

- Renamed "QueryError" to "TimeseriesNotFound". Since that's the only use-case, it's a little more accurate.
- Matched for the error in Nexus, translating it to `Ok(vec![])` with some explaining comments.

Reply:
That's excellent thanks. FWIW, this kind of thing will probably go away entirely in the future, since I think we'll eventually need to move the schema into CockroachDB. (See #1294)